third_party/move/move-vm/runtime/src/storage/ty_depth_checker.rs

```diff
@@ -0,0 +1,575 @@
+            | Type::U32
+            | Type::U64
+            | Type::U128
+            | Type::U256
+            | Type::Address
+            | Type::Signer => check_depth!(0),
+            Type::Reference(ty) | Type::MutableReference(ty) => self.check_depth_of_type_impl(
+                gas_meter,
+                traversal_context,
+                ty,
+                max_depth,
+                check_depth!(1),
+            )?,
+            Type::Vector(ty) => self.check_depth_of_type_impl(
+                gas_meter,
+                traversal_context,
+                ty,
+                max_depth,
+                check_depth!(1),
+            )?,
+            Type::Struct { idx, .. } => {
+                let formula =
+                    self.calculate_struct_depth_formula(gas_meter, traversal_context, idx)?;
+                check_depth!(formula.solve(&[]))
+            },
+            Type::StructInstantiation { idx, ty_args, .. } => {
+                let ty_arg_depths = ty_args
+                    .iter()
+                    .map(|ty| {
+                        self.check_depth_of_type_impl(
+                            gas_meter,
+                            traversal_context,
+                            ty,
+                            max_depth,
+                            check_depth!(0),
+                        )
+                    })
+                    .collect::<PartialVMResult<Vec<_>>>()?;
+                let formula =
+                    self.calculate_struct_depth_formula(gas_meter, traversal_context, idx)?;
+                check_depth!(formula.solve(&ty_arg_depths))
+            },
+            Type::Function { args, results, .. } => {
+                let mut ty_max_depth = depth;
+                for ty in args.iter().chain(results) {
+                    ty_max_depth = ty_max_depth.max(self.check_depth_of_type_impl(
+                        gas_meter,
+                        traversal_context,
+                        ty,
+                        max_depth,
+                        check_depth!(1),
+                    )?);
+                }
+                ty_max_depth
+            },
+            Type::TyParam(_) => {
+                return Err(
+                    PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+                        .with_message("Type parameter should be fully resolved".to_string()),
+                )
+            },
+        };
+
+        Ok(ty_depth)
+    }
+
+    /// Calculates the depth formula for a (possibly generic) struct or enum. Returns an error if
+    /// the struct definition is recursive: i.e., some struct A uses struct B that uses struct A.
+    fn calculate_struct_depth_formula(
+        &self,
+        gas_meter: &mut impl GasMeter,
+        traversal_context: &mut TraversalContext,
+        idx: &StructNameIndex,
+    ) -> PartialVMResult<DepthFormula> {
+        // If the struct is being visited, we found a recursive definition.
```

Comment:
Hm, let's say A{B, C} while internally B{C}. 
This should be okay but might trip up the pending formulas check, depending on the traversal order. But in principle, I don't see why we may not end up visiting C again and causing the error?

The tests also don't have such a no false positive case.

Reply:
Nice test case: added it as a test (only had generics because found them problematic with layouts).

Basically here is why this will be ok:

1. Check if struct is currently being visited, if not: use from cache or mark as pending, visit fields, unmark as pending.
2. While visiting fields, we will end up at some point not visiting any struct, and so terminate recursion and unmark pending adding the formula to cache. So visiting `C` is "mark(C), unmark(C), cache(C)"

For your example it means that we have:
```
mark(A)
  mark(B)
    mark(C)
    unmark(C)
    cache(C)
  unmark(B)
  cache(B)
  use_cached(C)
```

Reply:
Even without cache we are ok, because recomputing formula for C adds it to pending, but pops on exit (DFS-like)

Reply:
Traversal order is also part of the algorithm, otherwise gas charges are also different so we cannot just change the order without feature gating.